drivers/can/can_native_posix_linux.c

```diff
@@ -0,0 +1,429 @@
+{
+	struct can_npl_data *data = dev->data;
+	struct can_filter_context *filter_ctx;
+	int filter_id = -ENOSPC;
+
+	LOG_DBG("Setting filter ID: 0x%x, mask: 0x%x", filter->id,
+		    filter->id_mask);
+	LOG_DBG("Filter type: %s ID %s mask",
+		filter->id_type == CAN_STANDARD_IDENTIFIER ?
+				   "standard" : "extended",
+		((filter->id_type && (filter->id_mask == CAN_STD_ID_MASK)) ||
+		(!filter->id_type && (filter->id_mask == CAN_EXT_ID_MASK))) ?
+		"with" : "without");
+
+	k_mutex_lock(&data->filter_mutex, K_FOREVER);
+
+	for (int i = 0; i < ARRAY_SIZE(data->filters); i++) {
+		if (data->filters[i].rx_cb == NULL) {
+			filter_id = i;
+			break;
+		}
+	}
+
+	if (filter_id < 0) {
+		LOG_ERR("No free filter left");
+		k_mutex_unlock(&data->filter_mutex);
+		return filter_id;
+	}
+
+	filter_ctx = &data->filters[filter_id];
+	filter_ctx->rx_cb = cb;
+	filter_ctx->cb_arg = cb_arg;
+	filter_ctx->filter = *filter;
+
+	k_mutex_unlock(&data->filter_mutex);
+
+	LOG_DBG("Filter added. ID: %d", filter_id);
+
+	return filter_id;
+}
+
+static void can_npl_remove_rx_filter(const struct device *dev, int filter_id)
+{
+	struct can_npl_data *data = dev->data;
+
+	k_mutex_lock(&data->filter_mutex, K_FOREVER);
+	data->filters[filter_id].rx_cb = NULL;
+	k_mutex_unlock(&data->filter_mutex);
+
+	LOG_DBG("Filter removed. ID: %d", filter_id);
+}
+
+static int can_npl_get_capabilities(const struct device *dev, can_mode_t *cap)
+{
+	ARG_UNUSED(dev);
+
+	*cap = CAN_MODE_NORMAL | CAN_MODE_LOOPBACK;
+
+#if CONFIG_CAN_FD_MODE
+	*cap |= CAN_MODE_FD;
+#endif /* CONFIG_CAN_FD_MODE */
+
+	return 0;
+}
+
+static int can_npl_set_mode(const struct device *dev, can_mode_t mode)
+{
+	struct can_npl_data *data = dev->data;
+
+#ifdef CONFIG_CAN_FD_MODE
+	if ((mode & ~(CAN_MODE_LOOPBACK | CAN_MODE_FD)) != 0) {
+		LOG_ERR("unsupported mode: 0x%08x", mode);
+		return -ENOTSUP;
+	}
+#else
+	if ((mode & ~(CAN_MODE_LOOPBACK)) != 0) {
+		LOG_ERR("unsupported mode: 0x%08x", mode);
+		return -ENOTSUP;
+	}
+#endif /* CONFIG_CAN_FD_MODE */
+
+	data->loopback = (mode & CAN_MODE_LOOPBACK) != 0;
+	linux_socketcan_set_loopback(data->dev_fd, data->loopback);
+
+	data->mode_fd = (mode & CAN_MODE_FD) != 0;
+	linux_socketcan_set_mode_fd(data->dev_fd, data->mode_fd);
+
+	return 0;
+}
+
+static int can_npl_set_timing(const struct device *dev, const struct can_timing *timing)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(timing);
+
+	return 0;
+}
+
+#ifdef CONFIG_CAN_FD_MODE
+static int can_npl_set_timing_data(const struct device *dev, const struct can_timing *timing)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(timing);
+
+	return 0;
+}
+#endif /* CONFIG_CAN_FD_MODE */
+
+static int can_npl_get_state(const struct device *dev, enum can_state *state,
+			     struct can_bus_err_cnt *err_cnt)
+{
+	ARG_UNUSED(dev);
+
+	if (state != NULL) {
+		/* SocketCAN does not forward error frames by default */
+		*state = CAN_STATE_ERROR_ACTIVE;
+	}
+
+	if (err_cnt) {
+		err_cnt->tx_err_cnt = 0;
+		err_cnt->rx_err_cnt = 0;
+	}
+
+	return 0;
+}
```

Comment:
Forwarding of error frames can be enabled, though. It would be a nice addition, but we can leave It for later.

Reply:
Ok, I'll leave that for a later enhancement.